front/src/common/SpeedLimitByTagSelector/SpeedLimitByTagSelector.tsx

```diff
@@ -53,6 +52,8 @@ export default function SpeedLimitByTagSelector({
               dispatchUpdateSpeedLimitByTag(null);
             }
           }}
+          getOptionLabel={(option) => option}
+          getOptionValue={(option) => option}
```

Comment:
Still find pretty weird that we will have to do this each time we use this component

Reply:
Would seem to agree, but we have to accept an array of strings for the suggestions or have a specific format for them !

Reply:
The purpose of these two props is to allow users to pass their data in its raw form without needing to transform it beforehand. You provide the data as-is and pass two callbacks to specify how to display and identify each element in the list. The benefit becomes more apparent when dealing with more complex data than simple strings.

Maybe we could make these props optional if the list of select options consists of strings?

Reply:
I agree that allowing `<Select>` users to pass their own data type is worthwhile, and I also agree that for many cases it's cumbersome to pass trivial `getOptionLabel`/`getOptionValue` callbacks. I don't recommend adding more knobs to `<Select>` props because it would increase the complexity of its API. Instead, a separate helper function could be introduced. Here's one of the possible solutions:

```tsx
type SelectFixedOption = {
  label: string;
  value: string;
};

const withFixedOptions = (options: SelectFixedOption[]) => {
  return {
    options,
    getOptionLabel: (option) => option.label,
    getOptionValue: (option) => option.value,
  };
};

// usage:
const options = speedLimitsTagsList.map((tag) => ({ label: tag, value: tag }));
<Select {...withFixedOptions(options)} />
```

Many variations of this are possible as well, e.g. only providing `getOptionLabel`/`getOptionValue` helpers for `string[]` (but wouldn't work if you want to use a label different from the value, which is quite common):

```tsx
const withStringArray = (options: string[]) => {
  return {
    options,
    getOptionLabel: (option) => option,
    getOptionValue: (option) => option,
  };
};

// usage:
<Select {...withStringArray(speedLimitsTagsList)} />
```

or replacing `SelectOption.value` with an `onSelect` callback:

```tsx
type SelectFixedOption = {
  label: string;
  onSelect: () => void;
};

const withFixedOptions = (options: SelectFixedOption[]) => {
  const optionsMap = new Map(options.map((option, index) => [option, index.toString()]));
  return {
    options,
    getOptionLabel: (option) => option.label,
    getOptionValue: (option) => optionsMap.get(option),
    onChange: (option) => option?.onSelect(),
  };
};

// usage:
const options = speedLimitsTagsList.map((tag) =>
  ({ label: tag, onSelect: () => dispatchUpdateSpeedLimitByTag(tag) })
);
<Select {...withFixedOptions(options)} />
```

Reply:
TBH… I'm not even sure why `<Select>` requires callers to pass a `value` at all. I think we could just auto-generate a `value` for each option based on their index. We wouldn't loose the current selection on component re-renders (e.g. when the option list changes) because internally `<Select>` does `value={selectedOption ? getOptionValue(selectedOption) : undefined}`.

Reply:
I now realize that if we want to drop `value`, we wouldn't even need to generate them based on indexes: we can just grab the selected index via [selectedIndex](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/selectedIndex).

Reply:
i like the solution of having `withFixedOptions` but if have to change the internal functioning of the `Select`   component we can create an issue in` ui-core`

Reply:
> I agree that allowing `<Select>` users to pass their own data type is worthwhile, and I also agree that for many cases it's cumbersome to pass trivial `getOptionLabel`/`getOptionValue` callbacks. I don't recommend adding more knobs to `<Select>` props because it would increase the complexity of its API. Instead, a separate helper function could be introduced. Here's one of the possible solutions:
> 
> ```tsx
> type SelectFixedOption = {
>   label: string;
>   value: string;
> };
> 
> const withFixedOptions = (options: SelectFixedOption[]) => {
>   return {
>     options,
>     getOptionLabel: (option) => option.label,
>     getOptionValue: (option) => option.value,
>   };
> };
> 
> // usage:
> const options = speedLimitsTagsList.map((tag) => ({ label: tag, value: tag }));
> <Select {...withFixedOptions(options)} />
> ```
> 
> Many variations of this are possible as well, e.g. only providing `getOptionLabel`/`getOptionValue` helpers for `string[]` (but wouldn't work if you want to use a label different from the value, which is quite common):
> 
> ```tsx
> const withStringArray = (options: string[]) => {
>   return {
>     options,
>     getOptionLabel: (option) => option,
>     getOptionValue: (option) => option,
>   };
> };
> 
> // usage:
> <Select {...withStringArray(speedLimitsTagsList)} />
> ```
> 
> or replacing `SelectOption.value` with an `onSelect` callback:
> 
> ```tsx
> type SelectFixedOption = {
>   label: string;
>   onSelect: () => void;
> };
> 
> const withFixedOptions = (options: SelectFixedOption[]) => {
>   const optionsMap = new Map(options.map((option, index) => [option, index.toString()]));
>   return {
>     options,
>     getOptionLabel: (option) => option.label,
>     getOptionValue: (option) => optionsMap.get(option),
>     onChange: (option) => option?.onSelect(),
>   };
> };
> 
> // usage:
> const options = speedLimitsTagsList.map((tag) =>
>   ({ label: tag, onSelect: () => dispatchUpdateSpeedLimitByTag(tag) })
> );
> <Select {...withFixedOptions(options)} />
> ```

I totally agree with you. Thank you for your helpful feedback and examples!

Reply:
> I now realize that if we want to drop `value`, we wouldn't even need to generate them based on indexes: we can just grab the selected index via [`selectedIndex`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/selectedIndex).

I see your point about auto-generating values based on the index. However, I believe keeping the value prop is important. Without it, we wouldn’t be able to initialize the Select component with a value from a Redux store. This makes it harder to set the correct selected option when loading from the store.
Do you have any suggestions on how to handle this situation with your proposal?

Reply:
> we wouldn’t be able to initialize the Select component with a value from a Redux store

What's in the Redux store is completely orthogonal to the props passed to `Select` I believe. The current selected option in `<Select value=…>` is of type `T`, it's _not_ a `string` compared to `getOptionValue`.

In other words, I don't think dropping `getOptionValue` in `Select` would make any difference: in OSRD we would just drop it and everything would still continue to work as before.

See https://github.com/OpenRailAssociation/osrd-ui/pull/538